timemachine/cpp/src/barostat.cu

```diff
@@ -0,0 +1,267 @@
+
+#define AVOGADRO 6.0221367e23
+#define BOLTZ 0.008314462618
+
+namespace timemachine {
+
+MonteCarloBarostat::MonteCarloBarostat(
+    const int N,
+    const double pressure,  // Expected in Bar
+    const double temperature, // Kelvin
+    const std::vector<std::vector<int> > group_idxs,
+    const int interval,
+    const std::vector<BoundPotential *> bps,
+    const int seed) :
+    N_(N),
+    pressure_(pressure),
+    temperature_(temperature),
+    interval_(interval),
+    bps_(bps),
+    group_idxs_(group_idxs),
+    volume_scale_(0),
+    num_attempted_(0),
+    num_accepted_(0),
+    seed_(seed),
+    step_(0) {
+
+    // lets not have another facepalm moment again...
+    if(temperature < 100.0) {
+        std::cout << "warning temperature less than 100K" << std::endl;
+    }
+
+    if(pressure > 10.0) {
+        std::cout << "warning pressure more than 10bar" << std::endl;
+    }
+
+    mt_ = std::mt19937(seed_);
+    dist_ = std::uniform_real_distribution<double>(0.0, 1.0);
+
+    const int num_mols = group_idxs_.size();
+
+    gpuErrchk(cudaMalloc(&d_x_after_, N_*3*sizeof(*d_x_after_)));
+    gpuErrchk(cudaMalloc(&d_box_after_, 3*3*sizeof(*d_box_after_)));
+
+    gpuErrchk(cudaMalloc(&d_u_before_, N_*sizeof(*d_u_before_)));
+    gpuErrchk(cudaMalloc(&d_u_after_, N_*sizeof(*d_u_after_)));
+
+    gpuErrchk(cudaMalloc(&d_mol_idxs_, num_mols*sizeof(*d_mol_idxs_)));
+
+    int mol_idxs[num_mols];
+    for (int i = 0; i < num_mols; i++) {
+        std::vector<int> atoms = group_idxs[i];
+        // Sort the array in ascending order
+        std::sort(atoms.begin(), atoms.end());
+        mol_idxs[i] = atoms[0];
+    }
+    gpuErrchk(cudaMemcpy(d_mol_idxs_, mol_idxs, num_mols*sizeof(*d_mol_idxs_), cudaMemcpyHostToDevice));
+
+};
+
+MonteCarloBarostat::~MonteCarloBarostat() {
+    gpuErrchk(cudaFree(d_x_after_));
+    gpuErrchk(cudaFree(d_mol_idxs_));
+    gpuErrchk(cudaFree(d_box_after_));
+    gpuErrchk(cudaFree(d_u_before_));
+    gpuErrchk(cudaFree(d_u_after_));
+};
+
+
+void __global__ displace_atoms(
+    const int N, // Number of coords
+    double * __restrict__ coords, // Coordinates [N*3]
+    const double length_scale,
+    const double * __restrict__ box, // [9]
+    double * __restrict__ scaled_box, // [9]
+    const int num_molecules, // Number of molecules
+    const int * __restrict__ mol_idxs // [num_molecules]
+) {
+    const int mol_idx = blockIdx.x*blockDim.x + threadIdx.x;
+    if (mol_idx >= num_molecules) {
+        return;
+    }
+    double center_x = box[0*3+0] * 0.5;
+    double center_y = box[1*3+1] * 0.5;
+    double center_z = box[2*3+2] * 0.5;
+
+    const int starting_atom_idx = mol_idxs[mol_idx];
+    const int ending_atom_idx = mol_idx + 1 >= num_molecules ? N : mol_idxs[mol_idx + 1];
+    const int num_atoms = ending_atom_idx - starting_atom_idx;
+
+    double centroid_x = 0;
+    double centroid_y = 0;
+    double centroid_z = 0;
+    for (int i = starting_atom_idx; i < ending_atom_idx; i++) {
```

Comment:
This is actually an inefficient algorithm esp. for the thread that processes the protein group. It will block every other thread. Did you benchmark the perf on a protein system?

Reply:
After changes benchmarked per on protein vs water and it was the same. Great catch, should have noticed this